File: 5. Longest Palindromic Substring/Longest Palindromic Substring.go
package main

import "fmt"

func main() {
	var s string = "babad"
	result := longestPalindrome(s)
	fmt.Println(result)
}

func longestPalindrome(s string) string {
    var expan = func(l int, r int) string {
		for l >= 0 && r < len(s) && string(s[l]) == string(s[r]) {
			fmt.Println(l," dan ", r)
			l -= 1
			r += 1
		}
		var out_slice string = s[l+1:r]
		return  out_slice
	}

	var result string

	for i,_ := range s {
		var evenPalind string = expan(i,i)
		if len(evenPalind) > len(result) {
			result = evenPalind
		}

		var oddPalind string = expan(i,i+1)
		if len(oddPalind) > len(result) {
			result = oddPalind
		}
	}

	return result

}

File: 5. Longest Palindromic Substring/Longest Palindromic Substring.py
class Solution:
	@staticmethod
	def expan(l,r,s):
		while l >= 0 and r < len(s) and s[l] == s[r]:
			l -= 1
			r += 1
			print(l,r)
		
		# print(s[l+1],s[r])
		# print("+++++++++++++++++++")
		print(f"++> {s[l+1:r]}")
		return s[l+1:r]
    
	def longestPalindrome(self, s: str) -> str:
		result = []
		for i in range(len(s)):
			sub = self.expan(i,i,s)
			print(f"====> {sub}")
			if len(sub) > len(result):
				result = sub
				print(f"==== {result} ====")
			sub2 = self.expan(i,i+1,s)
			if len(sub2) > len(result):
				result = sub2
				print(f"==== {result} ====")
		
		return result

result = Solution().longestPalindrome("babad")
# result = Solution().longestPalindrome("cbbd")
print(result)
# a = "cbbd"
# print(a[-1],a[4])